### PR TITLE
implement GP-verification

### PR DESCRIPTION
**Pull request by vedi**

### CODE REVIEW

File: SoomlaAndroidExample/src/com/soomla/example/ExampleEventHandler.java [hunk 1]
```diff
@@ -20,9 +20,26 @@
 import android.os.Handler;
 import android.widget.Toast;
 import com.soomla.BusProvider;
-import com.soomla.SoomlaConfig;
 import com.soomla.SoomlaApp;
-import com.soomla.store.events.*;
+import com.soomla.SoomlaConfig;
```
Comment (refaelos) [inline SoomlaAndroidExample/src/com/soomla/example/ExampleEventHandler.java]: Why did you add all events here instead of `import com.soomla.store.events.*;` ?

File: SoomlaAndroidExample/AndroidManifest.xml [hunk 2]
```diff
@@ -1,13 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
           package="com.soomla.example"
-          android:versionCode="1"
-          android:versionName="1.0">
+          android:versionCode="4"
```
Comment (refaelos) [inline SoomlaAndroidExample/AndroidManifest.xml]: get these numbers back to 1 and 1.0

File: SoomlaAndroidExample/src/com/soomla/example/MuffinRushAssets.java [hunk 3]
```diff
@@ -194,7 +192,7 @@ public int getVersion() {
     public static final VirtualGood NO_ADS_GOOD = new LifetimeVG(
             "No Ads",                                                     // name
             "No More Ads!",                                                // description
-            "no_ads",                                                      // item id
+            NO_ADS_PRODUCT_ID,                                                      // item id
```
Comment (refaelos) [inline SoomlaAndroidExample/src/com/soomla/example/MuffinRushAssets.java]: get the value here back to "no_ads"

File: SoomlaAndroidExample/src/com/soomla/example/StoreExampleActivity.java [hunk 4]
```diff
@@ -99,8 +100,13 @@ construct the key at runtime from pieces or use bit manipulation (for example,
 
         Soomla.initialize("[CUSTOM SECRET HERE]");
         SoomlaStore.getInstance().initialize(storeAssets);
-        GooglePlayIabService.getInstance().setPublicKey("[YOUR PUBLIC KEY FROM THE MARKET]");
         GooglePlayIabService.AllowAndroidTestPurchases = true;
+        GooglePlayIabService iabService = GooglePlayIabService.getInstance();
+        iabService.setPublicKey("MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAgwyIs+GudcF/19zBscSeuaGaFNQ9ro3RfuwWOWAkbV54NMRtdZuTgvtUkDVO2221yZCSpTT4Kj55DX9/A7PguXyc1uGuy42Enzbp3d6mXESTnUucNi4jmxpikORxuVZHqW3/fZiaOcBWKb2frgRztkoGraKG9RIGve3E205oClbZQHkvc3Z8zi30WLkXeMbrA0HP3/2q2ZwF+z5kA4y2Lfv8LMsXRYM9hHTScsEIvRhPX8H3omRgECn4R+ZHh2WHjRw3KHnknbTLKsgFGsmsECQzPhjQTeC7eVfmDqhjT2yTJJSsYigRV2EAB6rG8qC+rivv18dtX95E0+6klUHU2QIDAQAB");
+        iabService.setVerifyPurchases(true);
```
Comment (refaelos) [inline SoomlaAndroidExample/src/com/soomla/example/StoreExampleActivity.java]: Are you sure you want this information publicly available ?
Remove verification from the example.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 5]
```diff
@@ -526,37 +522,22 @@ private void handleSuccessfulPurchase(IabPurchase purchase) {
...
-                // TODO: update on the issue in github
-                if (StoreInfo.isItemNonConsumable(pvi)) {
-                    if (StorageManager.getVirtualItemStorage(pvi).getBalance(pvi.getItemId()) == 1) {
-                        return;
+            case 0: {
+                if (mInAppBillingService instanceof IIabVerifiable) {
+                    IIabVerifiable iabVerifiable = (IIabVerifiable) mInAppBillingService;
+                    if (iabVerifiable.getVerifyPurchases()) {
+                        iabVerifiable.verifyPurchase(purchase, pvi);
                     }
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: there's a problem here ... in case `getVerifyPurchases` returns false nothing will happen and the purchase will hang forever.
You need to call `finalizeTransaction` on the purchase here.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 6]
```diff
@@ -526,37 +522,22 @@ private void handleSuccessfulPurchase(IabPurchase purchase) {
...
-
-                consumeIfConsumable(purchase, pvi);
-
                 break;
-
+            }
             case 1:
 
             case 2:
+                String developerPayload = purchase.getDeveloperPayload();
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: why are you saving developerPayload here?

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 7]
```diff
@@ -640,6 +621,51 @@ private void handleErrorResult(String message) {
         SoomlaUtils.LogError(TAG, "ERROR: SoomlaStore failure: " + message);
     }
 
+    private void finalizeTransaction(IabPurchase purchase, PurchasableVirtualItem pvi) {
+        SoomlaUtils.LogDebug(TAG, "IabPurchase successful.");
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: log: "IabPurchase successful. Finalizing transaction"

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 8]
```diff
@@ -640,6 +621,51 @@ private void handleErrorResult(String message) {
...
+
+        BusProvider.getInstance().post(new MarketPurchaseEvent(pvi, developerPayload, new HashMap<String, String>() {{
+                    put("token", token);
+                    put("orderId", orderId);
+                    put("originalJson", originalJson);
+                    put("signature", signature);
+                    put("userId", userId);
+                }}, null));
+
+        pvi.give(1);
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: I wanted to get your opinion on this a while ago... now it's a good time: Do you think it's smart to give the balance after the MarketPurchaseEvent is thrown? I'm just not sure about it.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 9]
```diff
@@ -640,6 +621,51 @@ private void handleErrorResult(String message) {
...
+        BusProvider.getInstance().post(new ItemPurchasedEvent(pvi.getItemId(), developerPayload));
+
+        consumeIfConsumable(purchase, pvi);
+    }
+
+    @Subscribe
+    public void on(MarketPurchaseVerificationEvent event) {
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: don't call the function `on`
use our naming convention for subscription functions. call it `onMarketPurchaseVerificationEvent`

Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: Btw, I'm a bit worried about this. It's an option for future bugs. What if this event will just be thrown somewhere in the future... SoomlaStore will always catch it.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java [hunk 10]
```diff
@@ -640,6 +621,51 @@ private void handleErrorResult(String message) {
...
+        if (event.isVerified()) {
+            this.finalizeTransaction(event.getPurchase(), event.getPvi());
+        } else {
+            SoomlaUtils.LogError(TAG, "Failed to verify transaction receipt. The user will not get what he just bought.");
+            BusProvider.getInstance().post(new UnexpectedStoreErrorEvent());
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: We have to get error codes ... developers should know this error is related to verification.
Let's put it as a future task.

File: SoomlaAndroidStore/src/com/soomla/store/StoreConfig.java [hunk 11]
```diff
@@ -33,4 +33,6 @@
      */
     public static final int METADATA_VERSION = 4;
 
+    public static final boolean VERIFY_PURCHASES = true;
```
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/StoreConfig.java]: Why do you have this? I thought you're using sharedpreferences.

File: SoomlaAndroidExample/src/com/soomla/example/ExampleEventHandler.java (hunk 1 above)
Comment (vedi) [inline SoomlaAndroidExample/src/com/soomla/example/ExampleEventHandler.java]: I have option "Optimize Import" turned on by default, and it rearrange it according my settings. Actually I tried to align these settings according Soomla's realities.

File: SoomlaAndroidExample/AndroidManifest.xml (hunk 2 above)
Comment (vedi) [inline SoomlaAndroidExample/AndroidManifest.xml]: done

File: SoomlaAndroidExample/src/com/soomla/example/MuffinRushAssets.java (hunk 3 above)
Comment (vedi) [inline SoomlaAndroidExample/src/com/soomla/example/MuffinRushAssets.java]: done

File: SoomlaAndroidExample/src/com/soomla/example/StoreExampleActivity.java (hunk 4 above)
Comment (vedi) [inline SoomlaAndroidExample/src/com/soomla/example/StoreExampleActivity.java]: I'll hide it

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 6 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: It's a legacy of step-by-step refactoring. Removed the redundant variable.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 7 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: done

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 8 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: I think it's ok if it's well documented. It depends on an use case, for instance if a developer needs to handle logic after market purchasing, but before changing the balance. Maybe it's the best point to implement Custom Server Verification, but in this case the developer should be able to change workflow for cases when validation failed.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 9 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: > don't call the function on
use our naming convention for subscription functions. call it onMarketPurchaseVerificationEvent

I found this way less error-prone. With this pattern you always have only handler for the class. If you call it `onMarketPurchaseVerificationEvent`, and eventually rename class to `ChangedMarketPurchaseVerificationEvent`, you can create another handler for this event by mistake.

Anyway, renamed.

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 10 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: do I have any action item here?

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 9 above)
Comment (vedi) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: > Btw, I'm a bit worried about this. It's an option for future bugs. What if this event will just be thrown somewhere in the future... SoomlaStore will always catch it.

I can create a separate event with `package` availability level, and fire it instead. What do you think?

File: SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java (hunk 10 above)
Comment (refaelos) [inline SoomlaAndroidStore/src/com/soomla/store/SoomlaStore.java]: Just add a trello ticket for error codes (and mention verification there)
